conclave-host/src/main/kotlin/com/r3/conclave/host/internal/attestation/EnclaveQuoteServiceGramineDCAP.kt

```diff
@@ -0,0 +1,22 @@
+package com.r3.conclave.host.internal.attestation
+
+import com.r3.conclave.common.internal.*
+import com.r3.conclave.host.internal.Native
+import com.r3.conclave.host.internal.NativeLoader
+import java.nio.ByteBuffer
+
+class EnclaveQuoteServiceGramineDCAP : EnclaveQuoteService() {
+
+    override fun initializeQuote(): Cursor<SgxTargetInfo, ByteBuffer> {
+        val targetInfo = Cursor.allocate(SgxTargetInfo)
+        //  When using Gramine, we do not need to load (dlsym) all the quoting libraries,
+        //    hence we pass a true bool value to avoid loading them.
+        Native.initQuoteDCAP(NativeLoader.libsPath.toString(), true, targetInfo.buffer.array())
+        return targetInfo
+    }
+
+    override fun retrieveQuote(report: ByteCursor<SgxReport>): ByteCursor<SgxSignedQuote> {
+        //  This is not executed in the context of Gramine flow.
+        throw IllegalStateException("The quote can't be retrieved from the host when running Gramine")
+    }
```

Comment:
This tells us where the abstraction no longer is fit for purpose. This `retrieveQuote` needs to be an implementation detail of GraalVM. We need an abstraction which is "higher level" to nicely fit Gramine into.

Reply:
There is some awkwardness here because of the how the call interface abstraction works. We have a fixed enum for all modes, but different calls are available (for example, the signed quote call is available in gramine despite not being needed). So as it stands we're pretty much bound to have some odds and ends like this.

I have some thoughts about how this can be dealt with, but it's a non trivial refactor.

Reply:
The `nicely` is the culprit here :-) Yes, with Gramine we break assumptions, specifically the fact that we have a fixed number of "handlers" to be registered and used.

We can surely refactor things around but I always wondering if we need continuous refactoring for every PR/feature we want to add, as this has always have impact to the speed of delivering features useful for users vs cleanness of the code, which is somehow not immediately needed (considering that, if things go well, we could even demise Graal at some point, and another round of refactoring will be needed).

In this case I think we can even get rid of this class for Gramine, as the initialization of the native libraries can be moved into the Gramine related attestation class.
But then we would need to have some conditional code around, in order to call `initializeQuote` in non-Gramine cases, which also mean that we would need to change the way we register the handlers, as in Gramine we now skip a couple of handlers (1 host 1 enclave).
We would also need to do a not small refactoring around, I think, for example `enclaveInterface` is mapped one to one with `enclaveHandle`, so they might need to be merged/improved.

We can fit everything in this PR, but then it would be more a big refactoring where Gramine attestation would become just a minor thing. I think we should be more granular in doing this massive changes, I'd suggest to find a compromise and postpone the refactoring in a separate PR.

Reply:
I think this is a pretty simple problem to fix at this stage as this is all still fresh in our memories. Not sure if it should be done in this PR though. Definitely best to keep this restricted to attestation.